src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -2,22 +2,134 @@
+import java.util.function.LongSupplier;
 
+@Slf4j
 public class TokenStatsTracker {
+
+
     private final Map<String, TraceContext> traceIdToContext = new ConcurrentHashMap<>();
 
-    public void startSpan(ProxyContext context) {
+    /**
+     * List of registered trackers in the cluster
+     */
+    private final Map<String, TrackerData> registeredTrackers = new ConcurrentHashMap<>();
+
+    private final RTopic topic;
+
+    private final Vertx vertx;
+
+    private final LongSupplier clock;
+
+    private final String trackerId;
+
+    public TokenStatsTracker(RedissonClient redis, Vertx vertx, LongSupplier clock) {
+        this.clock = clock;
+        this.vertx = vertx;
+        this.topic = redis.getTopic("deployment-cost-stats-event", new StringCodec());
+        topic.addListener(String.class, (channel, event) -> onEvent(event));
+        this.trackerId = UUID.randomUUID().toString();
+        vertx.setPeriodic(0, 60_000, event -> cleanupDeadTrackers());
+        vertx.setPeriodic(0, 10_000, event -> ping());
+    }
+
+    private void cleanupDeadTrackers() {
+        long currentTime = clock.getAsLong();
+        for (Map.Entry<String, TrackerData> entry : registeredTrackers.entrySet()) {
+            TrackerData trackerData = entry.getValue();
+            String trackerId = entry.getKey();
+            if (currentTime - trackerData.lastUpdateTs > 60_000) {
+                // complete all promises forcibly
+                trackerData.completePromises();
+                registeredTrackers.remove(trackerId);
+            }
+        }
+    }
+
+    private void ping() {
+        PingEvent event = new PingEvent(this.trackerId);
+        publish(Event.PING, event).onFailure(error -> log.error("Can't ping trackers due to error", error));
+    }
+
+    private void onEvent(String data) {
+        int index = data.indexOf(':');
+        String eventName = data.substring(0, index);
+        String json = data.substring(index + 1);
+        Event event = Event.valueOf(eventName);
+        switch (event) {
+            case PING -> handlePing(Objects.requireNonNull(ProxyUtil.convertToObject(json, PingEvent.class)));
+            case END_SPAN -> handleEndSpan(Objects.requireNonNull(ProxyUtil.convertToObject(json, EndSpanEvent.class)));
+            case START_SPAN -> handleStartSpan(Objects.requireNonNull(ProxyUtil.convertToObject(json, StartSpanEvent.class)));
+            default -> log.warn("Unsupported event {}", event);
+        }
+    }
+
+    private void handlePing(PingEvent event) {
+        if (this.trackerId.equals(event.trackerId)) {
+            return;
+        }
+        TrackerData trackerData = getTrackerData(event.trackerId);
+        trackerData.lastUpdateTs = clock.getAsLong();
+    }
+
+    private TrackerData getTrackerData(String trackerId) {
+        return registeredTrackers.computeIfAbsent(trackerId, key -> new TrackerData(clock.getAsLong()));
+    }
+
+    private void handleEndSpan(EndSpanEvent event) {
```

Comment:
please add javadoc/comment how this method behaves for caller (producer) and consumer (other core instances)

Reply:
fixed